File: src/fundamentals/validation.rs
use crate::*;
use std::rc::Rc;

#[derive(Copy, Clone, PartialEq, Eq)]
pub enum ValidationResult<T> {
    Valid(T),
    Invalid,
}

pub use ValidationResult::*;

pub type AnyValidationResult = ValidationResult<Any>;

#[derive(Clone)]
pub struct Validation<A: 'static + Clone, B: 'static>(
    Rc<dyn Fn(A, &mut Environment, &ProgramStack) -> Result<ValidationResult<B>>>,
);

impl<A: Clone, B> Validation<A, B> {
    pub fn new(
        validate: impl Fn(A, &mut Environment, &ProgramStack) -> Result<ValidationResult<B>> + 'static,
    ) -> Validation<A, B> {
        Validation(Rc::new(validate))
    }

    pub fn validate(
        &self,
        value: A,
        env: &mut Environment,
        stack: &ProgramStack,
    ) -> Result<ValidationResult<B>> {
        (self.0)(value, env, stack)
    }

    pub fn or(self, other: Validation<A, B>) -> Validation<A, B> {
        Validation::new(move |input: A, env, stack| {
            match self.validate(input.clone(), env, stack)? {
                Valid(new_value) => Ok(Valid(new_value)),
                Invalid => other.validate(input, env, stack),
            }
        })
    }
}

impl<A: Clone, B: Clone> Validation<A, B> {
    pub fn and<C>(self, other: Validation<B, C>) -> Validation<A, C> {
        Validation::new(
            move |input, env, stack| match self.validate(input, env, stack)? {
                Valid(new_value) => other.validate(new_value, env, stack),
                Invalid => Ok(Invalid),
            },
        )
    }

    pub fn join<C>(self, other: Validation<B, C>) -> Validation<A, (B, C)> {
        Validation::new(
            move |input, env, stack| match self.validate(input, env, stack)? {
                Valid(new_value) => match other.validate(new_value.clone(), env, stack)? {
                    Valid(result) => Ok(Valid((new_value, result))),
                    Invalid => Ok(Invalid),
                },
                Invalid => Ok(Invalid),
            },
        )
    }
}

pub type AnyValidation = Validation<Any, Any>;

impl AnyValidation {
    pub fn from<A: Clone, B: Clone>(validation: Validation<A, B>) -> AnyValidation {
        AnyValidation::new(move |value, env, stack| {
            let value = value.cast::<A>().clone();

            let result = validation.validate(value, env, stack)?;

            Ok(match result {
                Valid(v) => Valid(Any::from(v)),
                Invalid => Invalid,
            })
        })
    }
}

impl<A: Clone, B: Clone> Validation<A, B> {
    pub fn from_any(validation: AnyValidation) -> Validation<A, B> {
        Validation::new(move |value, env, stack| {
            let erased_result = validation.validate(Any::from(value), env, stack)?;

            let result = match erased_result {
                Valid(erased_value) => {
                    let value = erased_value.cast::<B>().clone();
                    Valid(value)
                }
                Invalid => Invalid,
            };

            Ok(result)
        })
    }
}

File: Sources/Fundamentals/Validation.swift
public enum ValidationResult<T> {
    case valid(T)
    case invalid
}

extension ValidationResult {
    public func map<U>(_ transform: (T) throws -> U) rethrows -> ValidationResult<U> {
        switch self {
        case .valid(let value):
            return try .valid(transform(value))
        case .invalid:
            return .invalid
        }
    }
    
    public func flatMap<U>(_ transform: (T) throws -> ValidationResult<U>) rethrows -> ValidationResult<U> {
        switch self {
        case .valid(let value):
            return try transform(value)
        case .invalid:
            return .invalid
        }
    }
}

public typealias AnyValidationResult = ValidationResult<Any>

public struct Validation<A, B> {
    public var validation: (A, inout Environment, ProgramStack) throws -> ValidationResult<B>
    public var debugLabel: String?
    
    public init(debugLabel: String? = nil, _ validation: @escaping (A, inout Environment, ProgramStack) throws -> ValidationResult<B>) {
        self.debugLabel = debugLabel
        self.validation = validation
    }
    
    public func callAsFunction(_ input: A, _ env: inout Environment, _ stack: ProgramStack) throws -> ValidationResult<B> {
        try self.validation(input, &env, stack)
    }
}

public func || <A, B>(
    lhs: Validation<A, B>,
    rhs: Validation<A, B>
) -> Validation<A, B> {
    Validation(debugLabel: "\(lhs.debugLabel ?? "<validation>") or \(rhs.debugLabel ?? "<validation>")") { a, env, stack in
        switch try lhs(a, &env, stack) {
        case .valid(let b):
            return .valid(b)
        case .invalid:
            return try rhs(a, &env, stack)
        }
    }
}

public func & <A, B, C>(
    lhs: Validation<A, B>,
    rhs: Validation<B, C>
) -> Validation<A, C> {
    Validation(debugLabel: "\(lhs.debugLabel ?? "<validation>") and \(rhs.debugLabel ?? "<validation>")") { a, env, stack in
        switch try lhs(a, &env, stack) {
        case .valid(let b):
            switch try rhs(b, &env, stack) {
            case .valid(let c):
                return .valid(c)
            case .invalid:
                return .invalid
            }
        case .invalid:
            return .invalid
        }
    }
}

public func && <A, B, C>(
    lhs: Validation<A, B>,
    rhs: Validation<B, C>
) -> Validation<A, (B, C)> {
    Validation(debugLabel: "\(lhs.debugLabel ?? "<validation>") with \(rhs.debugLabel ?? "<validation>")") { a, env, stack in
        switch try lhs(a, &env, stack) {
        case .valid(let b):
            switch try rhs(b, &env, stack) {
            case .valid(let c):
                return .valid((b, c))
            case .invalid:
                return .invalid
            }
        case .invalid:
            return .invalid
        }
    }
}

public typealias AnyValidation = Validation<Any, Any>

public func toAny<A, B>(_ validation: Validation<A, B>) -> AnyValidation {
    Validation(debugLabel: validation.debugLabel) { value, env, stack in
        switch try validation(value as! A, &env, stack) {
        case .valid(let newValue):
            return .valid(newValue)
        case .invalid:
            return .invalid
        }
    }
}

public func fromAny<A, B>(_ validation: AnyValidation) -> Validation<A, B> {
    Validation(debugLabel: validation.debugLabel) { value, env, stack in
        switch try validation(value, &env, stack) {
        case .valid(let newValue):
            return .valid(newValue as! B)
        case .invalid:
            return .invalid
        }
    }
}
